components/metris/internal/provider/azure/azure.go

```diff
@@ -103,93 +106,133 @@ func (a *Azure) Run(ctx context.Context) {
+		ctx, span = trace.StartSpan(ctx, "metris/provider/azure/processInstance")
+		defer span.End()
 
-					if instance.lastEvent == nil {
-						workerlogger.With("error", err).Error("could not get metrics, dropping events because no cached information")
-					} else {
-						workerlogger.With("error", err).Error("could not get metrics, using information from cache")
+		workerlogger = workerlogger.With("traceID", span.SpanContext().TraceID).With("spanID", span.SpanContext().SpanID)
+	}
 
-						eventData = instance.lastEvent
-					}
-				}
+	vmcaps := make(vmCapabilities)
 
-				if eventData != nil {
-					if err := a.sendMetrics(workerlogger, instance, eventData); err != nil {
-						workerlogger.With("error", err).Error("error parsing metric information, could not send event to EDP")
-					}
-				}
+	if obj, exists := a.vmCapsStorage.Get(instance.cluster.Region); exists {
+		if caps, ok := obj.(*vmCapabilities); ok {
+			vmcaps = *caps
+		}
+	} else {
+		workerlogger.Warnf("vm capabilities for region %s not found, some metrics won't be available", instance.cluster.Region)
+	}
 
-				// save changes to storage
-				a.instanceStorage.Put(instance.cluster.TechnicalID, instance)
+	var (
+		eventData       *EventData
+		err             error
+		errorOccurred   bool
+		rateLimited     bool
+		instanceDeleted bool
+	)
 
-				a.queue.Done(clusterid)
+	eventData, err = getMetrics(ctx, workerlogger, instance, &vmcaps, a.config.PollInterval, a.config.PollingDuration)
+
+	if err != nil {
+		errorOccurred = true
+		eventData, rateLimited, instanceDeleted = processError(ctx, workerlogger, instance, eventData, err, a.config.MaxRetries, a.instanceStorage)
+	}
 
-				// requeue item after X duration if client still in storage
-				if !a.queue.ShuttingDown() {
-					workerlogger.Debugf("requeuing cluster in %s", a.config.PollInterval)
-					a.queue.AddAfter(clusterid, a.config.PollInterval)
+	if !errorOccurred {
+		instance.retryAttempts = 0
+		a.instanceStorage.Put(instance.cluster.TechnicalID, instance)
+	}
+	if eventData != nil {
+		if err := a.sendMetrics(ctx, workerlogger, instance, eventData); err != nil {
+			workerlogger.With("error", err).Error("error parsing metric information, could not send eventData to EDP")
+		}
+		if !instanceDeleted {
+			a.instanceStorage.Put(instance.cluster.TechnicalID, instance)
+		}
+	}
+	return rateLimited
+}
+
+func processError(ctx context.Context, workerlogger log.Logger, instance *Instance, eventData *EventData, err error, maxRetries int, instanceStorage storage.Storage) (*EventData, bool, bool) {
+	var span *trace.Span
+	if tracing.IsEnabled() {
+		ctx, span = trace.StartSpan(ctx, "metris/provider/azure/processError")
+		defer span.End()
+	}
+	eventData = instance.lastEvent
+	workerlogger.With("error", err).Error("could not get metrics, using information from cache")
+	if eventData == nil {
+		workerlogger.With("error", err).Error("could not get metrics from cache, dropping events because no cached information")
```

Comment:
I believe this log can be moved to whoever called `processError`. Because the decision to act on nil `eventData` is done later.